packages/asset-swapper/src/utils/market_operation_utils/fills.ts

```diff
@@ -287,3 +297,29 @@ export function getPathAdjustedSlippage(
     const rateChange = maxRate.minus(totalRate);
     return rateChange.div(maxRate).toNumber();
 }
+
+export function getCompleteRate(
+    side: MarketOperation,
+    input: BigNumber,
+    output: BigNumber,
+    targetInput: BigNumber,
+): BigNumber {
+    if (input.eq(0) || output.eq(0) || targetInput.eq(0)) {
+        return ZERO_AMOUNT;
+    }
+    // Penalize paths that fall short of the entire input amount by a factor of
+    // input / targetInput => (i / t)
+    if (side === MarketOperation.Sell) {
+        // (o / i) * (i / t) => (o / t)
+        return output.div(targetInput);
+    }
+    // (i / o) * (i / t)
+    return input.div(output).times(input.div(targetInput));
```

Comment:
Bug fix. We were previously just doing `targetInput.div(output)` which made incomplete paths look really impressive, so we would make buy splits that made no sense.